File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int n = height.size();
        int left[n];
        int right[n];
        int max = -1;

        for(int i=0; i<n; i++){
            if(height[i] >= max ){
                max = height[i];
            }
            left[i] = max;
        }
        max = -1;

        for(int i=n-1; i>=0; i--){
            if(height[i] >= max){
                max = height[i];
            }
            right[i] = max;
        }
        int total = 0;
        for(int i=0; i<n; i++){
            total += min(left[i] , right[i]) - height[i];
        }
        return total;
    }
};

File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    def trap(self, height: list[int]) -> int:
        n = len(height)
        left =[0]*n
        right =[0]*n
        maxLen = -1

        for i in range(n):
            if height[i] >= maxLen:
                maxLen = height[i]
            left[i] = maxLen
        
        maxLen = -1

        for i in range(n-1 , -1 , -1):
            if height[i] >= maxLen:
                maxLen = height[i]
            right[i] = maxLen
        
        total = 0
        for i in range(n):
            total += min(left[i] , right[i]) - height[i]
        return total